### PR TITLE
Log strings removal from the binary

### LINKED ISSUE #42840: logging: v2: support removal of format strings in binary for dictionary logging
**Is your enhancement proposal related to a problem? Please describe.**
Currently, logging v2 requires the format strings to be in memory for packaging to work, as `cbprintf_package()` needs the format strings to process the argument lists. For dictionary logging, the format strings do not need to be present in memory and only representations of these strings are outputted (e.g. hashes or memory addresses). These representations are processed/retrieved offline using the source code or built binary and the actual format strings are retrieved. These allows the format strings to be stripped from the final binary to conserve space (with `CONFIG_LOG2_FMT_SECTION=y`). However, since `cbprintf_package()` requires these format strings to be in memory, they cannot be stripped from binary.

**Describe the solution you'd like**
With `CONFIG_LOG2_FMT_SECTION` enabled, the format strings can be stripped from the binary to conserve flash space to support dictionary logging.

**Describe alternatives you've considered**
None. The alternative is to keep the strings in binary, but it kind of defeats the purpose of using dictionary logging in the first place.

### LINKED ISSUE #44942: Dictionary logging to automatically remove strings from the installed image
I have been trying to use the dictionary-based logging feature because my target device has only 64KB ROM. My client creates a log in the flash which can be examined if a device is returned from the field. They currently fill half the space with strings used at run time, and with those same strings stored in the log. A dictionary solution which removes the strings entirely is highly desirable. This is what they assumed the Zephyr feature offers. It doesn't. It should. 

When I first looked at this, the strings could not be removed at all, so the feature only reduced the amount of data in the log. This is only half a solution. A recent merge means that the strings are now stored in a section called "log_strings_sections" which *can* be removed, presumably with objcopy. Sadly, that section is placed by the linker in the middle of the ROM. Removing it has no effect on the footprint of the installed image. 

A secondary issue was that removing the section with objcopy led to a warning about an "empty loadable section - was this your intention?", or words to that effect. That's disconcerting.

A tertiary issue is that there was little useful documentation about any of this potentially wonderful feature. There are just some comments that the feature exists, and an example program.

The simplest solution might be to reorder the standard linker script so that "log_strings_sections" is last in the ROM image, or not "loadable" (not sure of the right term), or outside of the physical ROM. The ELF contains a bunch of DWARF debug sections which are not installed on the device. Can we achieve something like that?

I wrote my own dictionary-based logger without much difficultly, but my client wants a Zephyr-only solution. I used a hash for the string IDs rather than an address/reference/offset (whatever Zephyr does). This meant that my dictionary section was simply optimised out of the image. Granted this is not super helpful if you want to read the dictionary, but I figured I could grab it from an unoptimised ELF or something - we'll see.

### PR DESCRIPTION
**Pull request by nordic-krch**
PR introduces option to remove logging strings from binary. It can be used if logging backend/frontend does not format the string but only uses the address of the string (dictionary based logging).
It is done by creating a memory region (devnull region) adjacent to the ROM region. Then section with logging strings is put in that region with `(COPY)` property so that it ends in elf file but not in binary.

Support is initially added to Cortex-M and RISCV but it should be fairly easy to extend other linker scripts.

`samples/subsys/logging/dictionary/sample.logger.basic.dictionary.uart_async_frontend` sample is now using that. 3200 bytes are saved by taking this approach:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       27860 B         1 MB      2.66%
             RAM:        6976 B       256 KB      2.66%
     DEVNULL_ROM:        3200 B       256 KB      1.22%
        IDT_LIST:          0 GB         2 KB      0.00%
```

Fixes #44942.
Fixes #42840.

### CODE REVIEW

File: include/zephyr/arch/arm/cortex_m/scripts/linker.ld [hunk 1]
```diff
@@ -35,7 +35,7 @@
 #if CONFIG_FLASH_LOAD_SIZE > 0
 #define ROM_SIZE CONFIG_FLASH_LOAD_SIZE
 #else
-#define ROM_SIZE (CONFIG_FLASH_SIZE*1K - CONFIG_FLASH_LOAD_OFFSET)
+#define ROM_SIZE (CONFIG_FLASH_SIZE * 1024 - CONFIG_FLASH_LOAD_OFFSET)
```
Comment (ithinuel) [inline include/zephyr/arch/arm/cortex_m/scripts/linker.ld:38]: Is that change required for this PR?

Comment (nordic-krch) [reply include/zephyr/arch/arm/cortex_m/scripts/linker.ld:38]: yes, because this value is used in preprocessor comparison (`#if ROM_SIZE > ...`) and preprocessor does not understand `1K` while linker does.

**Review by dcpleung — APPROVED** [review]

**Review by ithinuel — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
This cause build failure with `samples/subsys/logging/dictionary/sample.logger.basic.dictionary.uart_int_frontend` on `qemu_riscv32_smp`. See https://github.com/zephyrproject-rtos/zephyr/issues/64596.

**Review by Khalilheyrani6367 — APPROVED** [review]
